connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -338,17 +342,24 @@ private void processTargetStateChanges(Set<String> connectorTargetStateChanges)
     public void halt() {
         synchronized (this) {
             // Clean up any connectors and tasks that are still running.
-            log.info("Stopping connectors and tasks that are still assigned to the worker");
-            worker.stopConnectors();
-            worker.stopAndAwaitTasks();
+            log.info("Stopping connectors and tasks that are still assigned to this worker.");
+            List<Callable<Void>> callables = new ArrayList<>();
+            for (String connectorName : new ArrayList<>(worker.connectorNames())) {
+                callables.add(getConnectorCallable(connectorName, ConnectorStatus.State.DESTROYED));
+            }
+            for (ConnectorTaskId taskId : new ArrayList<>(worker.taskIds())) {
+                callables.add(getTaskCallable(taskId, TaskStatus.State.DESTROYED));
+            }
+            startAndStop(callables);
 
             member.stop();
 
             // Explicitly fail any outstanding requests so they actually get a response and get an understandable reason
             // for their failure
-            while (!requests.isEmpty()) {
-                HerderRequest request = requests.poll();
+            HerderRequest request = requests.pollFirst();
```

Comment:
Indeed, but in this case the two forms (`while` and `for`) are perfectly equivalent, containing exactly the same control statement as well as the same assignments. And since the `for` loop doesn't fit in one line and I'd have to break it, I think keeping the `while` loop form makes sense.